### ISSUE TITLE
Only run setup/teardown tasks if a test task needs to run (with build cache support)

### ISSUE THREAD

**Issue by rkrisztian** [gradle/gradle#12862]
### Context
We got Karma/Jasmine unit tests executed through Gradle with the [gradle-node-plugin](https://github.com/node-gradle/gradle-node-plugin), let's call it `npmTest` (executing command `npm test`). Instead of letting Karma run whatever browser it finds installed on the system (which may be incompatible), we also use the [gradle-docker-plugin](https://github.com/bmuschko/gradle-docker-plugin) to fire up a well-known instance of a browser, using e.g. the [elgalu/docker-selenium](https://github.com/elgalu/docker-selenium) images, which also let us record the screen in case of weird problems so we can debug easily. However, the docker plugin makes us map every single docker operation into a separate task, which was already problematic for multiple projects building in parallel (#10549). And now we got a new problem: those docker tasks are never up to date, and they keep running even when the `npmTest` state is `FROM-CACHE`.

I considered the idea of [a similar problem with `functionalTestRun`](https://discuss.gradle.org/t/how-can-i-start-a-server-conditionally-before-a-functionaltestrun/6389), but it is not gradle cache compatible, and in our case, the task is actually cacheable because they are unit tests. And it already feels wrong that we have no proper API to copy all up-to-date conditions to other tasks, and that all checks would be performed multiple times, let alone if we would cache the same results multiple times.

The reason I do the docker handling in Gradle is that we also need it for acceptance tests that also need a browser. The only difference there is that acceptance test tasks are never be up to date due to dependence on an external system (the backend).

### Expected Behavior
We need the possibility to depend on task(s) that will only get executed if the task that needs those dependent tasks is not up to date and has to be executed.

### Current Behavior
Only workarounds can be given right now and they are too hacky. Current solution even makes us consider dropping the gradle-docker-plugin, which is too bad. I do agree though that the granularity of docker tasks can be a bit too high with that plugin.

**Comment by CristianGM** [comment]
We have a lots of use cases like that, some with Docker, but others like running Android UI Tests on CI, it requires an emulator and I don't want to start an emulator if the tests are not going to run.

**Comment by rkrisztian** [comment]
Ping for keeping this issue alive. This comment has been updated 1 time(s).
